src/libfreeswan/rangetosubnet.3
.TH IPSEC_RANGETOSUBNET 3 "8 Sept 2000"
.SH NAME
ipsec rangetosubnet \- convert address range to subnet
.SH SYNOPSIS
.B "#include <freeswan.h>"
.sp
.B "const char *rangetosubnet(const ip_address *start,"
.ti +1c
.B "const ip_address *stop, ip_subnet *dst);"
.SH DESCRIPTION
.I Rangetosubnet
accepts two IP addresses which define an address range,
from
.I start
to
.I stop
inclusive,
and converts this to a subnet if possible.
The addresses must both be IPv4 or both be IPv6,
and the address family of the resulting subnet is the same.
.PP
.I Rangetosubnet
returns NULL for success and
a pointer to a string-literal error message for failure;
see DIAGNOSTICS.
.SH SEE ALSO
ipsec_initsubnet(3), ipsec_ttosubnet(3)
.SH DIAGNOSTICS
Fatal errors in
.I rangetosubnet
are:
mixed address families;
unknown address family;
.I start
and
.I stop
do not define a subnet.
.SH HISTORY
Written for the FreeS/WAN project by Henry Spencer.
.SH BUGS
The restriction of error reports to literal strings
(so that callers don't need to worry about freeing them or copying them)
does limit the precision of error reporting.
.PP
The error-reporting convention lends itself
to slightly obscure code,
because many readers will not think of NULL as signifying success.
A good way to make it clearer is to write something like:
.PP
.RS
.nf
.B "const char *error;"
.sp
.B "error = rangetosubnet( /* ... */ );"
.B "if (error != NULL) {"
.B "        /* something went wrong */"
.fi
.RE


NEWS
strongswan-4.3.1
----------------

- The nm plugin now passes DNS/NBNS server information to NetworkManager,
  allowing Gateway administrator to set DNS/NBNS configuration on clients
  dynamically.

strongswan-4.3.0
----------------

- Support for the IKEv2 Multiple Authentication Exchanges extension (RFC4739).
  Initiators and responders can use several authentication rounds (e.g. RSA
  followed by EAP) to authenticate. The new ipsec.conf leftauth/rightauth and
  leftauth2/rightauth2 parameters define own authentication rounds or setup
  constraints for the remote peer. See the ipsec.conf man page for more detials.

- If glibc printf hooks (register_printf_function) are not available,
  strongSwan can use the vstr string library to run on non-glibc systems.

- The IKEv2 charon daemon can now configure the ESP CAMELLIA-CBC cipher
  (esp=camellia128|192|256).

- Refactored the pluto and scepclient code to use basic functions (memory
  allocation, leak detective, chunk handling, printf_hooks, strongswan.conf
  attributes, ASN.1 parser, etc.) from the libstrongswan library.

- Up to two DNS and WINS servers to be sent via IKEv1 ModeConfig can be
  configured in the pluto section of strongswan.conf.


strongswan-4.2.14
-----------------

- The new server-side EAP RADIUS plugin (--enable-eap-radius)
  relays EAP messages to and from a RADIUS server. Succesfully
  tested with with a freeradius server using EAP-MD5 and EAP-SIM.

- A vulnerability in the Dead Peer Detection (RFC 3706) code was found by
  Gerd v. Egidy <gerd.von.egidy@intra2net.com> of Intra2net AG affecting
  all Openswan and strongSwan releases. A malicious (or expired ISAKMP)
  R_U_THERE or R_U_THERE_ACK Dead Peer Detection packet can cause the
  pluto IKE daemon to crash and restart. No authentication or encryption
  is required to trigger this bug. One spoofed UDP packet can cause the
  pluto IKE daemon to restart and be unresponsive for a few seconds while
  restarting. This DPD null state vulnerability has been officially
  registered as CVE-2009-0790 and is fixed by this release.

- ASN.1 to time_t conversion caused a time wrap-around for
  dates after Jan 18 03:14:07 UTC 2038 on 32-bit platforms.
  As a workaround such dates are set to the maximum representable
  time, i.e. Jan 19 03:14:07 UTC 2038.

- Distinguished Names containing wildcards (*) are not sent in the
  IDr payload anymore. 


strongswan-4.2.13
-----------------

- Fixed a use-after-free bug in the DPD timeout section of the
  IKEv1 pluto daemon which sporadically caused a segfault.

- Fixed a crash in the IKEv2 charon daemon occuring with
  mixed RAM-based and SQL-based virtual IP address pools. 

- Fixed ASN.1 parsing of algorithmIdentifier objects where the
  parameters field is optional.

- Ported nm plugin to NetworkManager 7.1.


strongswan-4.2.12
-----------------

- Support of the EAP-MSCHAPv2 protocol enabled by the option
  --enable-eap-mschapv2. Requires the MD4 hash algorithm enabled
  either by --enable-md4 or --enable-openssl.

- Assignment of up to two DNS and up to two WINS servers to peers via
  the IKEv2 Configuration Payload (CP). The IPv4 or IPv6 nameserver 
  addresses are defined in strongswan.conf.

- The strongSwan applet for the Gnome NetworkManager is now built and
  distributed as a separate tarball under the name NetworkManager-strongswan.

 
strongswan-4.2.11
-----------------

- Fixed ESP NULL encryption broken by the refactoring of keymat.c.
  Also introduced proper initialization and disposal of keying material.

- Fixed the missing listing of connection definitions in ipsec statusall
  broken by an unfortunate local variable overload.


strongswan-4.2.10
-----------------

- Several performance improvements to handle thousands of tunnels with almost
  linear upscaling. All relevant data structures have been replaced by faster
  counterparts with better lookup times.

- Better parallelization to run charon on multiple cores. Due to improved
  ressource locking and other optimizations the daemon can take full
  advantage of 16 or even more cores.

- The load-tester plugin can use a NULL Diffie-Hellman group and simulate
  unique identities and certificates by signing peer certificates using a CA
  on the fly.

- The redesigned stroke in-memory IP pool handles leases. The "ipsec leases"
  command queries assigned leases.

- Added support for smartcards in charon by using the ENGINE API provided by
  OpenSSL, based on patches by Michael Roßberg.

- The Padlock plugin supports the hardware RNG found on VIA CPUs to provide a
  reliable source of randomness.

strongswan-4.2.9
----------------

- Flexible configuration of logging subsystem allowing to log to multiple
  syslog facilities or to files using fine-grained log levels for each target.

- Load testing plugin to do stress testing of the IKEv2 daemon against self
  or another host. Found and fixed issues during tests in the multi-threaded
  use of the OpenSSL plugin.

- Added profiling code to synchronization primitives to find bottlenecks if
  running on multiple cores. Found and fixed an issue where parts of the
  Diffie-Hellman calculation acquired an exclusive lock. This greatly improves
  parallelization to multiple cores.

- updown script invocation has been separated into a plugin of its own to
  further slim down the daemon core.

- Separated IKE_SA/CHILD_SA key derivation process into a closed system,
  allowing future implementations to use a secured environment in e.g. kernel
  memory or hardware.

- The kernel interface of charon has been modularized. XFRM NETLINK (default)
  and PFKEY (--enable-kernel-pfkey) interface plugins for the native IPsec
  stack of the Linux 2.6 kernel as well as a PFKEY interface for the KLIPS
  IPsec stack (--enable-kernel-klips) are provided.

- Basic Mobile IPv6 support has been introduced, securing Binding Update
  messages as well as tunneled traffic between Mobile Node and Home Agent.

src/libfreeswan/rangetosubnet.c
/*
 * express an address range as a subnet (if possible)
 * Copyright (C) 2000  Henry Spencer.
 * 
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/lgpl.txt>.
 * 
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
 * License for more details.
 */
#include "internal.h"
#include "freeswan.h"

/*
 - rangetosubnet - turn an address range into a subnet, if possible
 *
 * A range which is a valid subnet will have a network part which is the
 * same in the from value and the to value, followed by a host part which
 * is all 0 in the from value and all 1 in the to value.
 */
err_t
rangetosubnet(from, to, dst)
const ip_address *from;
const ip_address *to;
ip_subnet *dst;
{
	unsigned const char *fp;
	unsigned const char *tp;
	unsigned fb;
	unsigned tb;
	unsigned const char *f;
	unsigned const char *t;
	size_t n;
	size_t n2;
	int i;
	int nnet;
	unsigned m;

	if (addrtypeof(from) != addrtypeof(to))
		return "mismatched address types";
	n = addrbytesptr(from, &fp);
	if (n == 0)
		return "unknown address type";
	n2 = addrbytesptr(to, &tp);
	if (n != n2)
		return "internal size mismatch in rangetosubnet";

	f = fp;
	t = tp;
	nnet = 0;
	for (i = n; i > 0 && *f == *t; i--, f++, t++)
		nnet += 8;
	if (i > 0 && !(*f == 0x00 && *t == 0xff)) {	/* mid-byte bdry. */
		fb = *f++;
		tb = *t++;
		i--;
		m = 0x80;
		while ((fb&m) == (tb&m)) {
			fb &= ~m;
			tb |= m;
			m >>= 1;
			nnet++;
		}
		if (fb != 0x00 || tb != 0xff)
			return "not a valid subnet";
	}
	for (; i > 0 && *f == 0x00 && *t == 0xff; i--, f++, t++)
		continue;

	if (i != 0)
		return "invalid subnet";

	return initsubnet(from, nnet, 'x', dst);
}



#ifdef RANGETOSUBNET_MAIN

#include <stdio.h>

void regress(void);

int
main(int argc, char *argv[])
{
	ip_address start;
	ip_address stop;
	ip_subnet sub;
	char buf[100];
	const char *oops;
	size_t n;
	int af;
	int i;

	if (argc == 2 && strcmp(argv[1], "-r") == 0) {
		regress();
		fprintf(stderr, "regress() returned?!?\n");
		exit(1);
	}

	if (argc < 3) {
		fprintf(stderr, "Usage: %s [-6] start stop\n", argv[0]);
		fprintf(stderr, "   or: %s -r\n", argv[0]);
		exit(2);
	}

	af = AF_INET;
	i = 1;
	if (strcmp(argv[i], "-6") == 0) {
		af = AF_INET6;
		i++;
	}

	oops = ttoaddr(argv[i], 0, af, &start);
	if (oops != NULL) {
		fprintf(stderr, "%s: start conversion failed: %s\n", argv[0], oops);
		exit(1);
	}
	oops = ttoaddr(argv[i+1], 0, af, &stop);
	if (oops != NULL) {
		fprintf(stderr, "%s: stop conversion failed: %s\n", argv[0], oops);
		exit(1);
	}
	oops = rangetosubnet(&start, &stop, &sub);
	if (oops != NULL) {
		fprintf(stderr, "%s: rangetosubnet failed: %s\n", argv[0], oops);
		exit(1);
	}
	n = subnettot(&sub, 0, buf, sizeof(buf));
	if (n > sizeof(buf)) {
		fprintf(stderr, "%s: reverse conversion", argv[0]);
		fprintf(stderr, " failed: need %ld bytes, have only %ld\n",
						(long)n, (long)sizeof(buf));
		exit(1);
	}
	printf("%s\n", buf);

	exit(0);
}

struct rtab {
	int family;
	char *start;
	char *stop;
	char *output;			/* NULL means error expected */
} rtab[] = {
	{4, "1.2.3.0",		"1.2.3.255",		"1.2.3.0/24"},
	{4, "1.2.3.0",		"1.2.3.7",		"1.2.3.0/29"},
	{4, "1.2.3.240",	"1.2.3.255",		"1.2.3.240/28"},
	{4, "0.0.0.0",		"255.255.255.255",	"0.0.0.0/0"},
	{4, "1.2.3.4",		"1.2.3.4",		"1.2.3.4/32"},
	{4, "1.2.3.0",		"1.2.3.254",		NULL},
	{4, "1.2.3.0",		"1.2.3.126",		NULL},
	{4, "1.2.3.0",		"1.2.3.125",		NULL},
	{4, "1.2.0.0",		"1.2.255.255",		"1.2.0.0/16"},
	{4, "1.2.0.0",		"1.2.0.255",		"1.2.0.0/24"},
	{4, "1.2.255.0",		"1.2.255.255",	"1.2.255.0/24"},
	{4, "1.2.255.0",		"1.2.254.255",	NULL},
	{4, "1.2.255.1",		"1.2.255.255",	NULL},
	{4, "1.2.0.1",		"1.2.255.255",		NULL},
	{6, "1:2:3:4:5:6:7:0",	"1:2:3:4:5:6:7:ffff",	"1:2:3:4:5:6:7:0/112"},
	{6, "1:2:3:4:5:6:7:0",	"1:2:3:4:5:6:7:fff",	"1:2:3:4:5:6:7:0/116"},
	{6, "1:2:3:4:5:6:7:f0",	"1:2:3:4:5:6:7:ff",	"1:2:3:4:5:6:7:f0/124"},
	{4, NULL,		NULL,			NULL},
};

void
regress()
{
	struct rtab *r;
	int status = 0;
	ip_address start;
	ip_address stop;
	ip_subnet sub;
	char buf[100];
	const char *oops;
	size_t n;
	int af;

	for (r = rtab; r->start != NULL; r++) {
		af = (r->family == 4) ? AF_INET : AF_INET6;
		oops = ttoaddr(r->start, 0, af, &start);
		if (oops != NULL) {
			printf("surprise failure converting `%s'\n", r->start);
			exit(1);
		}
		oops = ttoaddr(r->stop, 0, af, &stop);
		if (oops != NULL) {
			printf("surprise failure converting `%s'\n", r->stop);
			exit(1);
		}
		oops = rangetosubnet(&start, &stop, &sub);
		if (oops != NULL && r->output == NULL)
			{}		/* okay, error expected */
		else if (oops != NULL) {
			printf("`%s'-`%s' rangetosubnet failed: %s\n",
						r->start, r->stop, oops);
			status = 1;
		} else if (r->output == NULL) {
			printf("`%s'-`%s' rangetosubnet succeeded unexpectedly\n",
							r->start, r->stop);
			status = 1;
		} else {
			n = subnettot(&sub, 0, buf, sizeof(buf));
			if (n > sizeof(buf)) {
				printf("`%s'-`%s' subnettot failed:  need %ld\n",
						r->start, r->stop, (long)n);
				status = 1;
			} else if (strcmp(r->output, buf) != 0) {
				printf("`%s'-`%s' gave `%s', expected `%s'\n",
					r->start, r->stop, buf, r->output);
				status = 1;
			}
		}
	}
	exit(status);
}

#endif /* RANGETOSUBNET_MAIN */
